go/database/mpt/io/parallel_visit.go

```diff
@@ -0,0 +1,429 @@
+				if err := source.Close(); err != nil {
+					log.Printf("failed to close source: %v", err)
+				}
+			}
+		}
+	}()
+
+	// error in one of the sources interrupts processing
+	if err := errors.Join(openingErrors...); err != nil {
+		return err
+	}
+
+	for i := 0; i < NumWorker; i++ {
+		go func(source NodeSource) {
+			for !done.Load() {
+				// TODO: implement throttling
+				// get the next job
+				requestsMutex.Lock()
+				req, present := requests.Pop()
+				requestsMutex.Unlock()
+
+				// process the request
+				if !present {
+					continue
+				}
+
+				// fetch the node and put it into the responses
+				node, err := source.Get(req.id)
+
+				responsesMutex.Lock()
+				responses[req.id] = response{node, err}
+				responsesCond.Signal()
+				responsesMutex.Unlock()
+
+				// if there was a fetch error, stop the workers
+				if err != nil {
+					done.Store(true)
+					return
+				}
+
+				// derive child nodes to be fetched
+				switch node := node.(type) {
+				case *mpt.BranchNode:
+					children := node.GetChildren()
+					requestsMutex.Lock()
+					for i, child := range children {
+						id := child.Id()
+						if id.IsEmpty() {
+							continue
+						}
+						pos := req.position.Child(byte(i))
+						requests.Add(request{pos, child.Id()})
+					}
+					requestsMutex.Unlock()
+				case *mpt.ExtensionNode:
+					next := node.GetNext()
+					requestsMutex.Lock()
+					pos := req.position.Child(0)
+					requests.Add(request{pos, next.Id()})
+					requestsMutex.Unlock()
+				case *mpt.AccountNode:
+					if !cutAtAccounts {
+						storage := node.GetStorage()
+						id := storage.Id()
+						if !id.IsEmpty() {
+							requestsMutex.Lock()
+							pos := req.position.Child(0)
+							requests.Add(request{pos, id})
+							requestsMutex.Unlock()
+						}
+					}
+				}
+			}
+		}(sources[i])
+	}
+
+	// Perform depth-first iteration through the trie.
+	stack := []mpt.NodeId{root}
+	for len(stack) > 0 {
+		cur := stack[len(stack)-1]
+		stack = stack[:len(stack)-1]
+
+		var res response
+		responsesMutex.Lock()
+		for {
+			found := false
+			res, found = responses[cur]
+			if found {
+				delete(responses, cur)
+				break
+			}
+			responsesCond.Wait()
+		}
+		responsesMutex.Unlock()
+
+		if res.err != nil {
+			return res.err
+		}
+
+		switch visitor.Visit(res.node, mpt.NodeInfo{Id: cur}) {
+		case mpt.VisitResponseAbort:
+			return nil
+		case mpt.VisitResponsePrune:
+			continue
+		}
+
+		switch node := res.node.(type) {
+		case *mpt.BranchNode:
+			children := node.GetChildren()
+			for i := len(children) - 1; i >= 0; i-- {
+				id := children[i].Id()
+				if !id.IsEmpty() {
+					stack = append(stack, id)
+				}
+			}
+		case *mpt.ExtensionNode:
+			next := node.GetNext()
+			stack = append(stack, next.Id())
+		case *mpt.AccountNode:
+			if !cutAtAccounts {
+				storage := node.GetStorage()
+				id := storage.Id()
+				if !id.IsEmpty() {
+					stack = append(stack, id)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+type position struct {
```

Comment:
nit: needs a line of documentation; something like "position addresses a node within a tree by listing the path from the root node to the respective node" ...

Reply:
ok, will add